### ISSUE TITLE
[Feature Proposal] Add LoRA training support for pi0.5 (PyTorch backend)

### ISSUE THREAD

**Issue by sen-code-lost** [Physical-Intelligence/openpi#842]
**Motivation**
Currently, the PyTorch implementation in this repository seems to lack native LoRA training support. Many users (including myself) aim to fine-tune pi0.5 on consumer-grade hardware (e.g., RTX 4090), which is challenging with full parameter fine-tuning due to VRAM constraints.

**Proposal**
I have implemented a **native LoRA training pipeline** for the PyTorch version of pi0.5. This implementation is lightweight and efficient, enabling high-performance fine-tuning on a single GPU.

**Implementation Details**

*   **Zero Dependencies**: Unlike approaches using `peft` or other external libraries, this implementation is **pure PyTorch**. It does not bloat the project's dependency list and offers better integration with the existing codebase.
*   **Target Modules**: The implementation identifies and injects adapters into the Attention mechanism and MLP layers within the transformer blocks:
    *   *Attention*: `q_proj`, `k_proj`, `v_proj`, `o_proj`
    *   *FFN/MLP*: `gate_proj`, `up_proj`, `down_proj`
*   **Flexibility**: The implementation supports unfreezing specific components (e.g., Vision Encoder) alongside LoRA adapters.

**Parameter Breakdown**
Below is the configuration used for validation (Hybrid: LoRA for Transformer + Unfrozen Vision Encoder):

```text
Total Parameters:           3,643,299,600
Frozen Parameters:          3,202,149,376

Trainable Parameters:
-----------------------------------------
LoRA Adapters (Rank=16):       26,542,080  (0.73%)
Vision Encoder (Unfrozen):    412,442,352
Projection Layers:              2,165,792
-----------------------------------------
Total Trainable:              441,150,224  (12.11%)
```

**Results & Validation**
I have successfully trained a pi0.5 model on Libero dataset using this implementation.

*   **Hardware**: Single RTX 4090 (24GB)
*   **VRAM Usage**: ~16 GB (Peak during training)
*   **Training Steps**: 30k
*   **Batch Size**:  16
*   **Loss Curve(Libero Dataset)**:
    <img width="800" alt="Loss Curve" src="https://github.com/user-attachments/assets/4cabdb53-8b9d-484b-92ed-ad6604091a2b" />

*   **Evaluation (Libero Benchmark)**
    The fine-tuned model demonstrates promising results across Libero tasks:

| Model | Libero Spatial | Libero Object | Libero Goal | Libero 10 | **Average** |
| :--- | :---: | :---: | :---: | :---: | :---: |
| **π0.5 (PyTorch LoRA @ 30k)** | **96.6** | **98.2** | **97.8** | **95.2** | **96.95** |

**Next Steps**
I am ready to submit a Pull Request. Please let me know if this aligns with the roadmap and if there are any specific coding guidelines or branch preferences I should follow.

---

**Comment by s1ghhh** [comment]
awesome!!!

**Comment by s1ghhh** [comment]
Hi @sen-code-lost , thanks for your awsome work!
I'm trying to fine-tune pi0.5 with LoRA (pytorch). Have you submitted a PR for LoRA support, or is there any branch/fork I can use for LoRA fine-tuning?
Thanks a lot!

**Comment by sen-code-lost** [comment]
GPU memory usage: 

<img width="917" height="413" alt="Image" src="https://github.com/user-attachments/assets/b73955c5-7df9-4232-b7ed-a54a06f3216d" />

**Comment by sen-code-lost** [comment]
> Hi [@sen-code-lost](https://github.com/sen-code-lost) , thanks for your awsome work! I'm trying to fine-tune pi0.5 with LoRA (pytorch). Have you submitted a PR for LoRA support, or is there any branch/fork I can use for LoRA fine-tuning? Thanks a lot!

Hi! Yes, I've just submitted a PR for PyTorch LoRA support, you can track the progress here: #854.
If you want to try it out immediately before it gets merged, feel free to use my branch:
[https://github.com/sen-code-lost/openpi/tree/feat/pytorch-lora](https://www.google.com/url?sa=E&q=https%3A%2F%2Fgithub.com%2Fsen-code-lost%2Fopenpi%2Ftree%2Ffeat%2Fpytorch-lora)

**Comment by s1ghhh** [comment]
Great work, really appreciate your contribution!

**Comment by TIPTOEHIGHERZ** [comment]
Thanks for your work, I wonder which config was used for finetuning, thanks!

**Comment by MmMaiIIi** [comment]
awesome!!!

**Comment by F-Fer** [comment]
Love it!

**Comment by cotton-ahn** [comment]
Thanks a lot!

**Comment by rylinjames** [comment]
nice work getting pi0.5 lora down to ~16gb peak on a single 4090. the dep-free pure-pytorch approach (q/k/v/o + gate/up/down) is the right call for this repo, peft adds a lot for tooling you mostly don't need here. one thing worth double-checking if you haven't: with vision encoder unfrozen + lora on the transformer, norm stats on your action head need to be re-fit post-lora or you'll see clean train loss but bad real-world rollout (the lora shifts the action distribution slightly). reflex doesn't train, so this isn't our layer, but we do export pi0.5 checkpoints (including lora-merged) to onnx + fp16 trt for deploy with cos=+1.0 parity. if the deploy side of your lora checkpoints is interesting, github.com/rylinjames/reflex-vla has good-first-issues tagged, or dm romirj on discord / playindus@gmail.com for collaborator access + design-partner convo.

**Comment by giakhuyendihoc** [comment]
Awesome work! May I ask how long does it take you to finetune (30k step on libero) with the above config on 4090?

**Comment by LokeshBadisa** [comment]
Hi @sen-code-lost , is there a reason why you are full finetuning vision encoder instead of lora fine tuning it? Many thanks on your effort.
